### PR TITLE
[flytekit][1][SimpleTransformer] Binary IDL With MessagePack

### PR DESCRIPTION
**Pull request by Future-Outlier**
## Tracking issue
https://github.com/flyteorg/flyte/issues/5318

## Why are the changes needed?
These changes enable the `SimpleTransformer` to use `from_binary_idl` when receiving binary IDL inputs. This will be particularly useful when dealing with attribute access.

## What changes were proposed in this pull request?
1. Add the `from_binary_idl` method to the `TypeTransformer` class.
2. Use `MessagePackDecoder` to decode `msgpack bytes` into Python values.
3. add `_default_flytekit_decoder` in `MessagePackDecoder` to support future cases, such as `Dict[int, str]` (by setting `strict_map_key=False`).

## How was this patch tested?
Unit tests, local execution, and remote execution.

- local execution is tested by this PR: 
https://github.com/flyteorg/flytekit/pull/2751

- remote exeuction is tested by these 2 PRs:
https://github.com/flyteorg/flytekit/pull/2751
https://github.com/flyteorg/flyte/pull/5744

### Setup process
```python
from dataclasses import dataclass, field
from datetime import datetime, date, timedelta
from flytekit import workflow, task, ImageSpec

flytekit_hash = "c24077bce6e63bf8df0d80dbc2c5e2ff3322bca8"

flytekit = f"git+https://github.com/flyteorg/flytekit.git@{flytekit_hash}"

image = ImageSpec(
    packages=[flytekit],
    apt_packages=["git"],
    registry="localhost:30000",
)

@dataclass
class DC:
    a: int = 1
    b: float = 1.0
    c: bool = True
    d: str = "hello"
    e: datetime = field(default_factory=datetime.now)
    f: date = field(default_factory=date.today)
    g: timedelta = field(default_factory=lambda: timedelta(days=1))

@task(container_image=image)
def t_a(a: int):
    print(a)
    assert(type(a), int)

@task(container_image=image)
def t_b(b: float):
    print(b)
    assert(type(b), float)

@task(container_image=image)
def t_c(c: bool):
    print(c)
    assert(type(c) == bool)

@task(container_image=image)
def t_d(d: str):
    print(d)
    assert(type(d) == str)

@task(container_image=image)
def t_e(e: datetime):
    print(e)
    assert(type(e) == datetime)

@task(container_image=image)
def t_f(f: date):
    print(f)
    assert(type(f) == date)

@task(container_image=image)
def t_g(g: timedelta):
    print(g)
    assert(type(g) == timedelta)

@workflow
def dc_wf(dc: DC):
    t_a(dc.a)
    t_b(dc.b)
    t_c(dc.c)
    t_d(dc.d)
    t_e(dc.e)
    t_f(dc.f)
    t_g(dc.g)

if __name__ == "__main__":
    from flytekit.clis.sdk_in_container import pyflyte
    from click.testing import CliRunner
    import os

    runner = CliRunner()
    path = os.path.realpath(__file__)
    input_val = '{"a": 1, "b": 3.14, "c": true, "d": "hello"}'
    result = runner.invoke(pyflyte.main,
                           ["run", path, "dc_wf", "--dc", input_val])

    print("Local Execution: ", result.output)
    #
    result = runner.invoke(pyflyte.main,
                           ["run", "--remote", path,
                            "dc_wf", "--dc", input_val])
    print("Remote Execution: ", result.output)

```

### Screenshots
- local execution
image

- remote execution
image

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

### CODE REVIEW

File: flytekit/core/type_engine.py [hunk 1]
```diff
@@ -129,6 +130,8 @@ def __init__(self, name: str, t: Type[T], enable_type_assertions: bool = True):
         self._t = t
         self._name = name
         self._type_assertions_enabled = enable_type_assertions
+        self._msgpack_encoder: Dict[Type, MessagePackEncoder] = {}
```
Comment (Future-Outlier) [inline flytekit/core/type_engine.py:133]: The encoder is for the follow up PR, which will support `untyped dict`

File: flytekit/core/type_engine.py [hunk 2]
```diff
@@ -42,22 +44,21 @@
...
-)
+from flytekit.models.literals import Binary, Literal, LiteralCollection, LiteralMap, Primitive, Scalar, Union, Void
 from flytekit.models.types import LiteralType, SimpleType, TypeStructure, UnionType
 
 T = typing.TypeVar("T")
 DEFINITIONS = "definitions"
 TITLE = "title"
 
 
+# In Mashumaro, the default encoder uses strict_map_key=False, while the default decoder uses strict_map_key=True.
```
Comment (wild-endeavor) [inline flytekit/core/type_engine.py:55]: this function doesn't have anything to do with Mashumaro right?  why mention mashumaro in the comments?

File: flytekit/core/type_engine.py (hunk 1 above)
Comment (wild-endeavor) [reply flytekit/core/type_engine.py:133]: {} not good, delete

File: flytekit/core/type_engine.py [hunk 3]
```diff
@@ -221,6 +224,17 @@ def to_python_value(self, ctx: FlyteContext, lv: Literal, expected_python_type:
             f"Conversion to python value expected type {expected_python_type} from literal not implemented"
         )
 
+    def from_binary_idl(self, binary_idl_object: Binary, expected_python_type: Type[T]) -> Optional[T]:
+        if binary_idl_object.tag == "msgpack":
+            try:
+                decoder = self._msgpack_decoder[expected_python_type]
+            except KeyError:
+                decoder = MessagePackDecoder(expected_python_type, pre_decoder_func=_default_flytekit_decoder)
```
Comment (wild-endeavor) [inline flytekit/core/type_engine.py:232]: This part isn't clear to me... why are we using the MessagePackDecoder from Mashumaro?  isn't that just for dataclasses?

**Review by wild-endeavor — COMMENTED** [review]
could you help me understand the usage of mashumaro please?

File: flytekit/core/type_engine.py (hunk 2 above)
Comment (Future-Outlier) [reply flytekit/core/type_engine.py:55]: This function will be put into `_default_flytekit_decoder`.
For example, if we access `dataclasss.dict_int_str` in a workflow, then we will use `from_binary_idl` here to turn the Binary IDL object to `Dict[int, str]`.

Note: `dict_int_str` is `Dict[int, str]`.

**Pull request closed without merging**